cirq-google/cirq_google/engine/stream_manager_test.py

```diff
@@ -49,63 +50,109 @@
 
     async def quantum_run_stream(
         self, requests: AsyncIterator[quantum.QuantumRunStreamRequest], **kwargs
     ) -> Awaitable[AsyncIterable[quantum.QuantumRunStreamResponse]]:
         """Fakes the QuantumRunStream RPC.
 
-        Expects the number of requests to be the same as len(self.responses_and_exceptions).
-
-        For every request, a response or exception is popped from `self.responses_and_exceptions`.
-        Before the next request:
-            * If it is a response, it is sent back through the stream.
-            * If it is an exception, the exception is raised.
+        Once a request is received, it is appended to `stream_requests`, and the test calling
+        `wait_for_request()` is notified.
 
-        This fake does not support out-of-order responses.
+        The response is sent when a test calls `reply()` with a `QuantumRunStreamResponse`. If a
+        test calls `reply()` with an exception, it is raised here to the `quantum_run_stream()`
+        caller.
 
-        No responses are ever made if `self.responses_and_exceptions` is empty.
+        This is called from the asyncio thread.
         """
+        responses_and_exceptions: asyncio.Queue = asyncio.Queue()
+        self._responses_and_exceptions_future.try_set_result(responses_and_exceptions)
 
-        async def run_async_iterator():
+        async def read_requests():
             async for request in requests:
                 self.stream_requests.append(request)
-
-                if not self.responses_and_exceptions:
-                    while True:
-                        await asyncio.sleep(1)
-
-                response_or_exception = self.responses_and_exceptions.pop(0)
-                if isinstance(response_or_exception, BaseException):
+                self._request_ready.add(None)
+
+        async def response_iterator():
+            asyncio.create_task(read_requests())
+            while True:
+                response_or_exception = await responses_and_exceptions.get()
+                if isinstance(response_or_exception, quantum.QuantumRunStreamResponse):
+                    yield response_or_exception
+                else:  # isinstance(response_or_exception, BaseException)
+                    self._responses_and_exceptions_future = duet.AwaitableFuture[asyncio.Queue]()
                     raise response_or_exception
-                response_or_exception.message_id = request.message_id
-                yield response_or_exception
 
         await asyncio.sleep(0)
-        return run_async_iterator()
+        return response_iterator()
 
     async def cancel_quantum_job(self, request: quantum.CancelQuantumJobRequest) -> None:
+        """Records the cancellation in `cancel_requests`.
+
+        This is called from the asyncio thread.
+        """
         self.cancel_requests.append(request)
         await asyncio.sleep(0)
 
+    async def wait_for_requests(self, num_requests=1):
```

Comment:
I'd suggest returning the received requests here, which would allow tests to verify properties of the requests or tailor the response to the request if needed.

Reply:
Hmm, verifying properties of requests can be achieved by inspecting `stream_requests`, as it does now at the end of every test case, and I'm not sure if tailoring the response to the request is desirable because ideally unit tests should ideally assert the exact expected behavior. Adding this could also potentially make it confusing when to use the returned requests here vs `stream_requests`.

WDYT?

Reply:
I guess it seems more flexible to at least have the option of checking messages as you go, rather than only waiting until the end and looking at `stream_requests`. I realize there may be different philosphies of how tests should be structured, but in my view it's good to fail early as soon as a problem is detected, rather than defer all checks to the end when it may be harder to tell where exactly things went wrong.

Reply:
The internal style guide recommends keeping the "arrange", "act", and "assert" sections distinct (presumably for verification clarity). But I updated the method to make it flexible 👍